packages/hardhat-core/src/internal/hardhat-network/provider/fork/ForkBlockchain.ts

```diff
@@ -277,10 +298,17 @@ export class ForkBlockchain implements PBlockchain {
     ) {
       return undefined;
     }
-    const transaction = new Transaction(rpcToTxData(rpcTransaction), {
+
+    const chainId = this._common.chainId();
+    const txData = rpcToTxData(rpcTransaction);
+    const senderPubKey = new Transaction(txData).getSenderPublicKey();
```

Comment:
I don't think this would work. If the tx also uses eip155 this will fail, as it's default to a mainnet common.

Reply:
We should get the sender pub key from txData and the chainId.

Reply:
It should look something like this:

```ts
const msgHash = this.hash(false);
const v = bufferToInt(this.v);

return ecrecover(
  msgHash,
  v,
  this.r,
  this.s,
  this. _implementsEIP155() ? this.getChainId() : undefined,
);
```